### PR TITLE
Use OAuth CAS Flow

### PR DESCRIPTION
**Pull request by tataihono**
This PR primary goal is to **add new OAuth support without breaking existing flows**.

On the way I added in a few bonus features which will make this package easier to use:
- apply JSLinting styles to files
- refactor configuration while still supporting old method
- refactor StartAuthentication method to reduce code duplication
- refactor errors to support events rather than callbacks while still supporting old method
- add a README to speed up on-boarding for developers using this package
- add in some smart defaults for Cru developers

@Omicron7 Please code-review. Thanks.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      'service': 'service',
+      'token': 'token/new'
+  },
+  ticketUrl: '',
+  maxAttempts: 3,
+  requireAccessToken: false,
+  cacheAccessToken: false,
+  cacheExpiresMinutes: 25,
+  managedApis: [],
+  casBaseUrl: 'https://thekey.me',
```
Comment (frett) [inline README.md]: I would probably use these defaults:

```
casBaseUrl: 'https://thekey.me/cas/' ,
casLoginPath: 'login',
casTicketPath: 'api/oauth/ticket',
```

it's a java webapp that happens to be currently deployed in the `cas` subdirectory.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+2. module checks URL fragment and cache for access_token (none found)
+3. module redirects user to The Key using the implicit OAuth grant
+4. user logs in and The Key redirects user back to redirect url in module configuration with access_token(The Key) in URL fragment
+5. app makes a request to a managedApi without a session
+6. module checks URL fragment and cache for access_token (found)
+7. module contacts The Key (casTicketPath) using access_token: access_token(The Key) and service: managedApi being requested
+8. Key returns a ticket to module
+9. module uses ticket to authenticate with auth-api(authenticationApiBaseUrl)
+10. auth-api returns access_token(service) to module
+11. original request sent with access_token(rails) to service
```
Comment (frett) [inline README.md]: `access_token(service)` to follow the name of step 10

**Comment by frett** [comment]
There is also an OAuth attributes endpoint available, not sure if that would be beneficial to include a method to access it using the access_token or not.

`https://thekey.me/cas/api/oauth/attributes`

we also provide an `expires_in` parameter in the implicit response which could be used for cache duration.

and I really need to document what OAuth errors can be returned from the endpoints

File: bower.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "angular-cas-auth-api",
-  "version": "0.1.3",
+  "version": "0.1.4",
   "authors": [
     "Brian Zoetewey <brian.zoetewey@ccci.org>"
```
Comment (Omicron7) [inline bower.json]: You should probably add yourself here too @ardation

File: dist/cas-auth-api.js [hunk 4]
```diff
@@ -1,285 +1,449 @@
...
                 // invalidate access token and cache
-                _accessToken = undefined;
-                if (_cacheAccessToken) {
+                config.accessToken = undefined;
+                if (config.cacheAccessToken && angular.isDefined(lscache)) {
                     lscache.remove('access_token');
                 }
 
+                if (config.oAuth === true && getAccessTokenFromHash().access_token === undefined) {
+                    $injector.get('$window').location.href =
```
Comment (Omicron7) [inline dist/cas-auth-api.js:331]: Does this handle encoding any invalid characters in the url, if say clientId or redirectUrl have characters that need to be escaped?

**Comment by Omicron7** [comment]
Looks good to me. :+1:

File: bower.json (hunk 3 above)
Comment (tataihono) [inline bower.json]: yay :+1:

File: dist/cas-auth-api.js (hunk 4 above)
Comment (tataihono) [inline dist/cas-auth-api.js:331]: this should be fixed with a5e37c7. Added support for `encodeURIComponent` of the clientId and redirectUrl.

**Comment by tataihono** [comment]
@frett Oh yeah that would be cool functionality we could add. Added to ToDos.

**Pull request merged**